### PR TITLE
Make sure we transmit the actual debug port to next dev mode run

### LINKED ISSUE #40848: devmode breaks when updating pom.xml
### Describe the bug

in past I could add a maven dependency and quakrus devmode would restart.

now when I do it devmode crashes.

```
ERROR: transport error 202: bind failed: Address already in use
ERROR: JDWP Transport dt_socket failed to initialize, TRANSPORT_INIT(510)
JDWP exit error AGENT_ERROR_TRANSPORT_INIT(197): No transports initialized [src/jdk.jdwp.agent/share/native/libjdwp/debugInit.c:700]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:57 min
[INFO] Finished at: 2024-05-26T23:38:51+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:3.10.2:dev (default-cli) on project jconqbot: Failed to run: Dev mode process did not complete successfully -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
```

### Expected behavior

_No response_

### Actual behavior

_No response_

### How to Reproduce?

_No response_

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by gsmet**
When starting dev mode, the debug port might not be free and we might use a random one.
When restarting dev mode (for instance when you adjust the pom.xml), we should try to reuse this port instead of defaulting to the initially non free one.

Also the port being free is tested when we build `DevModeRunner`, so we need to make sure we stop the old runner before build()ing the new instance (and not just before we start the new one).

Fixes #40848

### CODE REVIEW

File: devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java [hunk 1]
```diff
@@ -478,15 +478,19 @@ public void close() throws IOException {
                     }
                     if (!changed.isEmpty()) {
                         getLog().info("Changes detected to " + changed + ", restarting dev mode");
+
+                        // stop the runner before we build the new one as the debug port being free
+                        // is tested when building the runner
+                        runner.stop();
```
Comment (geoand) [inline devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:484]: Do we need to do this in Gradle as well?

Comment (gsmet) [reply devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:484]: Your guess is as good as mine :). I will try to have a look.

Comment (gsmet) [reply devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java:484]: I don't see the restart logic in Gradle but I'm not very familiar with how it works.

Maybe @glefloch or @aloubyansky know more.

But I wouldn't block this PR as the current Maven behavior is a bit annoying.

**Review by geoand — APPROVED** [review]

**Comment by aloubyansky** [comment]
We don't support restarts on Gradle build script changes atm

**Pull request merged**